core/drivers/crypto/caam/hash/caam_hash.c

```diff
@@ -0,0 +1,643 @@
+	},
+	{
+		/* sha224 */
+		.type = OP_ALGO(SHA224),
+		.size_digest = TEE_SHA224_HASH_SIZE,
+		.size_block = TEE_MAX_HASH_SIZE,
+		.size_ctx = HASH_MSG_LEN + TEE_SHA256_HASH_SIZE,
+	},
+	{
+		/* sha256 */
+		.type = OP_ALGO(SHA256),
+		.size_digest = TEE_SHA256_HASH_SIZE,
+		.size_block = TEE_MAX_HASH_SIZE,
+		.size_ctx = HASH_MSG_LEN + TEE_SHA256_HASH_SIZE,
+	},
+	{
+		/* sha384 */
+		.type = OP_ALGO(SHA384),
+		.size_digest = TEE_SHA384_HASH_SIZE,
+		.size_block = TEE_MAX_HASH_SIZE * 2,
+		.size_ctx = HASH_MSG_LEN + TEE_SHA512_HASH_SIZE,
+	},
+	{
+		/* sha512 */
+		.type = OP_ALGO(SHA512),
+		.size_digest = TEE_SHA512_HASH_SIZE,
+		.size_block = TEE_MAX_HASH_SIZE * 2,
+		.size_ctx = HASH_MSG_LEN + TEE_SHA512_HASH_SIZE,
+	},
+};
+
+/*
+ * Maximum number of entry in the descriptor
+ */
+#define MAX_DESC_ENTRIES 20
+
+/*
+ * Full hashing data SW context
+ */
+struct hashdata {
+	uint32_t *descriptor; /* Job descriptor */
+	struct caamblock blockbuf; /* Temporary Block buffer */
+	struct caambuf ctx; /* Hash Context used by the CAAM */
+	const struct hashalg *alg; /* Reference to the algo constants */
+};
+
+/*
+ * Format the HASH context to keep the reference to the
+ * operation driver
+ */
+struct crypto_hash {
+	struct crypto_hash_ctx hash_ctx; /* Crypto Hash API context */
+	struct hashdata *ctx; /* Hash Context */
+};
+
+/*
+ * Keep the HW hash limit because after the initialization
+ * of the module, we don't have the CAAM Controller base address
+ * to call the function returning the HW capacity.
+ */
+static uint8_t caam_hash_limit;
+
+/*
+ * Returns the reference to the driver context
+ *
+ * @ctx  API Context
+ */
+static struct crypto_hash *to_hash_ctx(struct crypto_hash_ctx *ctx)
+{
+	assert(ctx && ctx->ops == &hash_ops);
+
+	return container_of(ctx, struct crypto_hash, hash_ctx);
+}
+
+/*
+ * Free the internal hashing data context
+ *
+ * @ctx    [in/out] Caller context variable
+ */
+static void do_free_intern(struct hashdata *ctx)
+{
+	HASH_TRACE("Free Context (0x%" PRIxPTR ")", (uintptr_t)ctx);
+
+	if (ctx) {
+		/* Free the descriptor */
+		caam_free_desc(&ctx->descriptor);
+
+		/* Free the Temporary buffer */
+		caam_free_buf(&ctx->blockbuf.buf);
+
+		/* Free the context register */
+		caam_free_buf(&ctx->ctx);
+	}
+}
+
+/*
+ * Allocate the internal hashing data context
+ *
+ * @ctx    [in/out] Caller context variable
+ */
+static enum CAAM_Status do_allocate_intern(struct hashdata *ctx)
+{
+	TEE_Result ret = CAAM_OUT_MEMORY;
+
+	HASH_TRACE("Allocate Context (0x%" PRIxPTR ")", (uintptr_t)ctx);
+
+	/* Allocate the descriptor */
+	ctx->descriptor = caam_alloc_desc(MAX_DESC_ENTRIES);
+	if (!ctx->descriptor) {
+		HASH_TRACE("Allocation descriptor error");
+		goto exit_alloc;
+	}
+
+	/* Initialize the block buffer */
+	ctx->blockbuf.filled = 0;
+	ctx->blockbuf.max = ctx->alg->size_block;
+
+	/* Allocate the CAAM Context register */
+	ret = caam_alloc_align_buf(&ctx->ctx, ctx->alg->size_ctx);
+
+#ifdef HASH_DEBUG
```

Comment:
Isn't it enough to select debugging with `HASH_TRACE()`?

Reply:
no because if HASH_DEBUG is not define the macro HASH_TRACE is nothing and so behavior is not the same. In plus code is not going to build.
Anyway, change a bit the if to remove the HASH_DEBUG check.